rviz_simulator/src/camera.cpp

```diff
@@ -0,0 +1,497 @@
+
+  double x_double_prime = x_prime * coefficient + 2 * camera_properties_.p1 * x_prime * y_prime +
+                          camera_properties_.p2 * (r_raise_2 + 2 * x_prime_squared);
+  double y_double_prime = y_prime * coefficient + camera_properties_.p1 * (r_raise_2 + 2 * y_prime_squared) +
+                          2 * camera_properties_.p2 * x_prime * y_prime;
+
+  double u = camera_properties_.fx * x_double_prime + camera_properties_.cx;
+  double v = camera_properties_.fy * y_double_prime + camera_properties_.cy;
+
+  Eigen::Vector2d pixelCoordinates;
+  pixelCoordinates[0] = u;
+  pixelCoordinates[1] = v;
+
+  return pixelCoordinates;
+}
+
+bool Camera::isWithinImage(Eigen::Vector2d pixelCoords)
+{
+  if (pixelCoords.x() >= 0 && pixelCoords.x() < this->camera_properties_.image_width && pixelCoords.y() >= 0 &&
+      pixelCoords.y() < this->camera_properties_.image_height)
+  {
+    ROS_INFO_STREAM("Valid corner coordinate\n");
+    return true;
+  }
+  ROS_INFO_STREAM("Invalid corner coordinate\n");
+  return false;
+}
+
+bool Camera::isInFrontOfCamera(Eigen::Affine3d camera_T_target)
+{
+  if (camera_T_target.translation().z() > 0)
+  {
+    ROS_INFO_STREAM("Target in front of camera\n");
+    return true;
+  }
+  ROS_INFO_STREAM("Target behind camera\n");
+  return false;
+}
+
+bool Camera::isInFrontOfTarget(Eigen::Affine3d target_T_camera)
+{
+  if (target_T_camera.translation().y() > 0)
+  {
+    ROS_INFO_STREAM("Camera in front of target\n");
+    return true;
+  }
+  ROS_INFO_STREAM("Camera behind target\n");
+  return false;
+}
+
+bool Camera::isInRange(std::vector<Eigen::Vector2d> corners)
+{
+  std::vector<Eigen::Vector2d> differences;
+  differences.push_back(corners[0] - corners[1]);
+  differences.push_back(corners[1] - corners[2]);
+  differences.push_back(corners[2] - corners[3]);
+  differences.push_back(corners[3] - corners[0]);
+  differences.push_back(corners[1] - corners[3]);
+  differences.push_back(corners[0] - corners[2]);
+
+  for (int i = 0; i < differences.size(); i++)
+  {
+    if (differences[i].norm() < this->camera_properties_.min_distance_between_target_corners)
+    {
+      ROS_INFO_STREAM("Target not in range\n.");
+      return false;
+    }
+  }
+  ROS_INFO_STREAM("Target in range\n.");
+  return true;
+}
+
+void Camera::dumpCameraPropertiesToYAMLFile(std::string output_file_name)
+{
+  YAML::Emitter out;
+  out << YAML::BeginMap;  // level 0 map
+
+  out << YAML::Key << "image_width" << YAML::Value << this->camera_properties_.image_width;
+  out << YAML::Key << "image_height" << YAML::Value << this->camera_properties_.image_height;
+  out << YAML::Key << "camera_name" << YAML::Value << this->camera_properties_.camera_name;
+
+  // camera matrix
+  out << YAML::Key << "camera_matrix" << YAML::Value;
+  out << YAML::BeginMap;  // camera matrix map
+  out << YAML::Key << "rows" << YAML::Value << 3;
+  out << YAML::Key << "cols" << YAML::Value << 3;
+  out << YAML::Key << "data" << YAML::Value;
```

Comment:
Does not conform to format specified in the design document.

Reply:
Fixed.